src/vacancies/vacancies/spiders/vacancies.py

```diff
@@ -0,0 +1,304 @@
+import json
+import random
+from typing import Any, Generator, List, Optional
+
+from scrapy.http import TextResponse
+from scrapy.spiders import Spider
+
+from vacancies.items import (
+    Address,
+    AreaType,
+    Employer,
+    IdName,
+    KeySkill,
+    Metro,
+    Specialization,
+    VacancyItem,
+)
+
+
+class VacanciesSpider(Spider):
+    """
+    Парсинг вакансий с api.hh.ru.
+    """
+
+    name = "vacancies"
+    start_urls = ["https://api.hh.ru/areas"]
+
+    def parse(self, response: TextResponse, **kwargs: Any) -> Optional[Generator]:
+        """
+        Парсинг данных о регионах.
+
+        :param response: Ответ от сервера.
+        :param kwargs:
+        :return:
+        """
+
+        json_res = json.loads(response.body)
+        if not isinstance(json_res, list) or len(json_res) < 1:
+            return None
+
+        areas_ru: List = []
+        area_ru_id = "113"  # Russia
+        for item in json_res:
+            if item["id"] == area_ru_id:
+                areas_ru = item["areas"]
+                break
+
+        if not areas_ru:
+            return None
+
+        areas_to_parse = list({area["id"] for area in areas_ru})
+        random.shuffle(areas_to_parse)
+        self.logger.info(f"Area IDs count – {len(areas_to_parse)}.")
+
+        # &industry=7&specialization=1
+        # page = random.randint(0, 10)
+        for area_id in areas_to_parse:
+            next_url = f"https://api.hh.ru/vacancies?per_page=100&area={area_id}"
+            yield response.follow(next_url, callback=self.parse_pages)
+
+        return None
+
+    def parse_pages(self, response: TextResponse, **kwargs: Any) -> Optional[Generator]:
+        """
+        Парсинг списка вакансий в указанном регионе с пагинацией.
+
+        :param response: Ответ от сервера.
+        :param kwargs:
+        :return:
+        """
+
+        json_res = json.loads(response.body)
```

Comment:
same question as in the previous method, can we get anything non-deserializable here and what will happen in such a situation?

Reply:
For `json.loads` case I've added special handling.